lib/services/role.go

```diff
@@ -541,43 +995,72 @@ func (set RoleSet) CheckLogins(ttl time.Duration) ([]string, error) {
 	return out, nil
 }
 
-// CheckAccessToServer checks if role set has access to server based
-// on combined role's selector and attempted login
+// CheckAccessToServer checks if a role has access to a node. Deny rules are
+// checked first then allow rules. Access to a node is determined by
+// namespaces, labels, and logins.
 func (set RoleSet) CheckAccessToServer(login string, s Server) error {
 	var errs []error
 
+	// check deny rules first: a single matching namespace, label, or login from
+	// the deny role set prohibits access.
 	for _, role := range set {
-		matchNamespace := MatchNamespace(role.GetNamespaces(), s.GetNamespace())
-		matchLabels := MatchLabels(role.GetNodeLabels(), s.GetAllLabels())
-		matchLogin := MatchLogin(role.GetLogins(), login)
+		matchNamespace := MatchNamespace(role.GetNamespaces(Deny), s.GetNamespace())
+		matchLabels := MatchLabels(role.GetNodeLabels(Deny), s.GetAllLabels())
+		matchLogin := MatchLogin(role.GetLogins(Deny), login)
+		if matchNamespace || matchLabels || matchLogin {
+			errorMessage := fmt.Sprintf("Role %v denied access to node %v: deny rule matched; match(namespace=%v, label=%v, login=%v)",
+				role.GetName(), s.GetHostname(), matchNamespace, matchLabels, matchLogin)
+			log.Warnf("[RBAC] " + errorMessage)
+			return trace.AccessDenied(errorMessage)
```

Comment:
this should be lowercase still. Warning can be upper case, but error should be lower case